api/include/opentelemetry/context/threadlocal_context.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include "opentelemetry/context/context.h"
+#include "opentelemetry/context/runtime_context.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+// The ThreadLocalContext class is a derived class from RuntimeContext and
+// provides a wrapper for propogating context through cpp thread locally.
+// This file must be included to use the RuntimeContext class if another
+// implementation has not been registered.
+class ThreadLocalContext : public RuntimeContext
+{
+public:
+  ThreadLocalContext() = default;
+
+  // Return the current context.
+  Context *InternalGetCurrent() override { return stack_.Top(); }
+
+  // Resets the context to a previous value stored in the
+  // passed in token. Returns true if successful, false otherwise
+  bool InternalDetach(Token &token) override
+  {
+    if (stack_.Top() == nullptr)
+    {
+      return false;
+    }
+    if (!(token == *stack_.Top()))
+    {
+      return false;
+    }
+    stack_.Pop();
+    return true;
+  }
+
+  // Sets the current 'Context' object. Returns a token
+  // that can be used to reset to the previous Context.
+  Token InternalAttach(Context *context) override
+  {
+    stack_.Push(context);
+    Token old_context = CreateToken(context);
+    return old_context;
+  }
+
+private:
+  // A nested class to store the attached contexts in a stack.
+  class Stack
+  {
+    friend class ThreadLocalContext;
+
+    Stack()
+    {
+      size_     = 0;
+      capacity_ = 0;
+      base_     = nullptr;
+    }
+
+    // Pops the top Context* off the stack and returns it.
+    Context *Pop()
+    {
+      if (size_ <= 0)
+      {
+        return nullptr;
+      }
+      int index = size_ - 1;
+      size_--;
+      return base_[index];
+    }
+
+    // Returns the Context* at the top of the stack.
+    Context *Top()
+    {
+      if (size_ <= 0)
+      {
+        return nullptr;
+      }
+      return base_[size_ - 1];
+    }
+
+    // Pushes the passed in context pointer to the top of the stack
+    // and resizes if necessary.
+    void Push(Context *context)
+    {
+      size_++;
+      if (size_ > capacity_)
+      {
+        Resize(size_ * 2);
+      }
+      base_[size_ - 1] = context;
+    }
+
+    // Reallocates the storage array to the pass in new capacity size.
+    void Resize(int new_capacity)
+    {
+      int old_size = size_ - 1;
+      if (new_capacity == 0)
+      {
+        new_capacity = 2;
+      }
+      Context **temp = new Context *[new_capacity];
+      memcpy(temp, base_, sizeof(Context *) * old_size);
+      delete[] base_;
+      base_ = temp;
+    }
+
+    ~Stack() { delete[] base_; }
+
+    int size_;
+    int capacity_;
+    Context **base_;
+  };
+
+  static thread_local Stack stack_;
+};
+thread_local ThreadLocalContext::Stack ThreadLocalContext::stack_ = ThreadLocalContext::Stack();
```

Comment:
Is it part of this PR to provide proper context resolution across thread? The current approach will initialize `stack_` with an empty stack for each new thread, so we won't be able to preserve parent context information.

Reply:
My thinking is that inheriting context across thread should be explicit. This is what's in my mind https://github.com/census-instrumentation/opencensus-python/blob/a4496ddaf59ffa3ec7cda806e56dcc320bf2d2df/context/opencensus-context/examples/explicit_threading.py#L43.

Reply:
So should we add that explicit propagation to this PR?

Reply:
I don't this it is under this PR's scope 😄

Reply:
Got it, makes sense.